counsel.el

```diff
@@ -820,14 +834,18 @@ Optional INITIAL-INPUT is the initial input in the minibuffer."
   (let* ((cands obarray)
          (pred 'commandp)
          (sort t))
-    (when (require 'smex nil 'noerror)
-      (unless smex-initialized-p
-        (smex-initialize))
-      (when (smex-detect-new-commands)
-        (smex-update))
-      (setq cands smex-ido-cache)
-      (setq pred nil)
-      (setq sort nil))
+    (cond (smex-initialized-p
+           (when (smex-detect-new-commands)
+             (smex-update))
+           (setq cands smex-ido-cache)
+           (setq pred nil)
+           (setq sort nil))
+          (amx-initialized
+           (when (amx-detect-new-commands)
+             (amx-update))
+           (setq cands amx-cache)
+           (setq pred nil)
+           (setq sort nil)))
```

Comment:
Perhaps the `cond` should be optimised for `amx`, seeing as it is the more actively maintained package?

Reply:
I wanted to keep the preference for smex as it is the one in the documentation. But I agree. If so, then it should be indicated in the documentation.

Reply:
The [suggested diff](https://github.com/abo-abo/swiper/pull/1580#issuecomment-390140345) extends the docstring of `counsel-M-x` to mention `amx`/`smex` integration. AIUI, their precedence shouldn't matter that much to users, as they are unlikely to have both installed, but there's no harm in saying as much in the docstring.